### PR TITLE
Add Support for subsectioned Values

### LINKED ISSUE #193: Add support for sections and custom introduction
Add support for the section just like bitnami readme generator where parameters can be broken down into sub-parameters using @section keyword 
https://github.com/bitnami/charts/blob/main/bitnami/redis/README.md

This will be helpful to separate DB sections, common values sections etc

### PR DESCRIPTION
**Pull request by Haepaxlog**
This PR closes #193 by adding section info to necessary structs and also retrieving
this section mapping through `# @section -- [section name]` annotations.

In order to keep compatibility with regular unsectioned `values.yaml` files, this means also
introducing a separate Sections field to `chartTemplateData` , through which a template
similar to `valuesTableTemplate` called `valuesTableSectionedTemplate` can be evoked to generate
files similar to this:
```markdown
## Values

### General
| Key | Type | Default | Description |
|-----|------|---------|-------------|
| controller.extraVolumes[0].name | string | `"config-volume"` |  |
| controller.image.repository | string | `"nginx-ingress-controller"` |  |
| controller.image.tag | string | `"18.0831"` |  |
| controller.name | string | `"controller"` |  |
| controller.service.annotations."external-dns.alpha.kubernetes.io/hostname" | string | `"stupidchess.jmn23.com"` | Hostname to be assigned to the ELB for the service |
| controller.service.type | string | `"LoadBalancer"` |  |

### Some Section
| Key | Type | Default | Description |
|-----|------|---------|-------------|
| controller.extraVolumes[0].configMap.name | string | `"nginx-ingress-config"` | Uses the name of the configmap created by this chart |
| controller.persistentVolumeClaims | list | the chart will construct this list internally unless specified | List of persistent volume claims to create. |
| controller.podLabels | object | `{}` | The labels to be applied to instances of the controller pod |

### Special Attention
| Key | Type | Default | Description |
|-----|------|---------|-------------|
| controller.ingressClass | string | `"nginx"` | You can also specify value comments like this |
| controller.publishService | object | `{"enabled":false}` | This is a publishService |
| controller.replicas | int | `nil` | Number of nginx-ingress pods to load balance between |
```
or HTML data:

<h3>General</h3>
<table>
	<thead>
		<th>Key</th>
		<th>Type</th>
		<th>Default</th>
		<th>Description</th>
	</thead>
	<tbody>
		<tr>
			<td>controller.extraVolumes[0].name</td>
			<td>string</td>
			<td><pre lang="json">
"config-volume"
</pre>
</td>
			<td></td>
		</tr>
		<tr>
			<td>controller.image.repository</td>
			<td>string</td>
			<td><pre lang="json">
"nginx-ingress-controller"
</pre>
</td>
			<td></td>
		</tr>
		<tr>
			<td>controller.image.tag</td>
			<td>string</td>
			<td><pre lang="json">
"18.0831"
</pre>
</td>
			<td></td>
		</tr>
		<tr>
			<td>controller.name</td>
			<td>string</td>
			<td><pre lang="json">
"controller"
</pre>
</td>
			<td></td>
		</tr>
		<tr>
			<td>controller.service.annotations."external-dns.alpha.kubernetes.io/hostname"</td>
			<td>string</td>
			<td><pre lang="json">
"stupidchess.jmn23.com"
</pre>
</td>
			<td>Hostname to be assigned to the ELB for the service</td>
		</tr>
		<tr>
			<td>controller.service.type</td>
			<td>string</td>
			<td><pre lang="json">
"LoadBalancer"
</pre>
</td>
			<td></td>
		</tr>
	</tbody>
</table>
<h3>Some Section</h3>
<table>
	<thead>
		<th>Key</th>
		<th>Type</th>
		<th>Default</th>
		<th>Description</th>
	</thead>
	<tbody>
		<tr>
			<td>controller.extraVolumes[0].configMap.name</td>
			<td>string</td>
			<td><pre lang="json">
"nginx-ingress-config"
</pre>
</td>
			<td>Uses the name of the configmap created by this chart</td>
		</tr>
		<tr>
			<td>controller.persistentVolumeClaims</td>
			<td>list</td>
			<td><pre lang="">
the chart will construct this list internally unless specified
</pre>
</td>
			<td>List of persistent volume claims to create.</td>
		</tr>
		<tr>
			<td>controller.podLabels</td>
			<td>object</td>
			<td><pre lang="json">
{}
</pre>
</td>
			<td>The labels to be applied to instances of the controller pod</td>
		</tr>
	</tbody>
</table>
<h3>Special Attention</h3>
<table>
	<thead>
		<th>Key</th>
		<th>Type</th>
		<th>Default</th>
		<th>Description</th>
	</thead>
	<tbody>
		<tr>
			<td>controller.ingressClass</td>
			<td>string</td>
			<td><pre lang="json">
"nginx"
</pre>
</td>
			<td>You can also specify value comments like this</td>
		</tr>
		<tr>
			<td>controller.publishService</td>
			<td>object</td>
			<td><pre lang="json">
{
  "enabled": false
}
</pre>
</td>
			<td>This is a publishService</td>
		</tr>
		<tr>
			<td>controller.replicas</td>
			<td>int</td>
			<td><pre lang="json">
null
</pre>
</td>
			<td>Number of nginx-ingress pods to load balance between</td>
		</tr>
	</tbody>
</table>

I'd be happy if this feature makes it in in some shape or form as I can see the need for documentation people to have it integrated into the tool :)

### CODE REVIEW

**Comment by klaudiagrz** [comment]
@Nepo26 any chances of implementing this feature in the nearest future? Looks super helpful!

**Comment by Nepo26** [comment]
Hi @klaudiagrz,

It really seems promising, I just need to review it, but I'm with little time recently. I can't promise anything, however I'll try to focus some time in this weekend.

**Comment by Nepo26** [comment]
Got some time now. I'm looking into it.

**Comment by kenrick-tan-partior** [comment]
Hey @Nepo26, did you manage to look at this PR yet? :)

File: pkg/document/model.go [hunk 1]
```diff
@@ -75,6 +82,50 @@ func sortValueRows(valueRows []valueRow) {
...
+func sortSectionedValueRows(sectionedValueRows []section) {
+	sortOrder := viper.GetString("sort-values-order")
+
+	if sortOrder != FileSortOrder && sortOrder != AlphaNumSortOrder {
+		log.Warnf("Invalid sort order provided %s, defaulting to %s", sortOrder, AlphaNumSortOrder)
+		sortOrder = AlphaNumSortOrder
+	}
+
+	for _, section := range sectionedValueRows {
+		sort.Slice(section.SectionItems, func(i, j int) bool {
```
Comment (norwoodj) [inline pkg/document/model.go:94]: This whole sort section looks identical to the one used in `sortValueRows`. Could you factor that out into a standalone function that takes a `[]valueRow` and a `sortOrder` and uses it?

File: pkg/document/model.go [hunk 2]
```diff
@@ -92,6 +143,39 @@ func getUnsortedValueRows(document *yaml.Node, descriptions map[string]helm.Char
 	return createValueRowsFromField("", nil, document.Content[0], descriptions, true)
 }
 
+func getUnsortedSectionedValueRows(valueRows []valueRow) []section {
+	var valueRowsSectionSorted []section
```
Comment (norwoodj) [inline pkg/document/model.go:147]: This variable name makes it sound like they're sorted :wink:

File: pkg/document/template.go [hunk 3]
```diff
@@ -274,6 +274,84 @@ func getValuesTableTemplates() string {
 	return valuesSectionBuilder.String()
 }
 
+func getValuesTableSectionedTemplates() string {
```
Comment (norwoodj) [inline pkg/document/template.go:277]: I don't like the approach here. I don't think that the addition of sections should necessitate adding new templates. Instead, I propose the following:

- The existing values table templates will change to render the `Sectioned` templates you've implemented here
- If no section is specified for a value then it belongs to a _default_ section.
- If no value specifies a section, then no section header is rendered for the _default_ section, e.g. the values will look exactly the same as they do today.
- If one ore more values specifies a section, the sections are rendered as you have done here, and any values that do not specify a section will be rendered in a section at the end, titled `Other Values` or something generic like that.

Does this all make sense?

File: pkg/helm/chart_info.go [hunk 4]
```diff
@@ -268,9 +270,10 @@ func parseChartValuesFileComments(chartDirectory string, values *yaml.Node, lint
 		// If we've already found a values comment, on the next line try and parse a custom default value. If we find one
```
Comment (norwoodj) [inline pkg/helm/chart_info.go:270]: Update this comment to reflect that we can also continue and parse a section comment now. Add a note about the continuation regex match as well, that should have been done before.

**Comment by norwoodj** [comment]
Hey @Haepaxlog I'm sorry for the slow response on this. I really like the idea here, and I've added some comments that I think would improve things slightly. Thanks so much for the contribution, if you make the couple changes I've proposed, happy to merge it.

**Review by norwoodj — CHANGES_REQUESTED** [review]

File: pkg/document/model.go (hunk 2 above)
Comment (Haepaxlog) [reply pkg/document/model.go:147]: True, naming things is hard :D

File: pkg/document/template.go (hunk 3 above)
Comment (Haepaxlog) [reply pkg/document/template.go]: Makes sense, and also seems to be a better approach here

**Comment by ksputo** [comment]
hey @norwoodj , could you please have a look at the changes? Thank you 🙇

File: pkg/document/template.go [hunk 5]
```diff
@@ -208,12 +208,33 @@ func getValuesTableTemplates() string {
 	valuesSectionBuilder.WriteString(`{{ define "chart.valuesHeader" }}## Values{{ end }}`)
 
 	valuesSectionBuilder.WriteString(`{{ define "chart.valuesTable" }}`)
+	valuesSectionBuilder.WriteString("{{ if .Sections.Sections }}")
+	valuesSectionBuilder.WriteString("{{ range .Sections.Sections }}")
+	valuesSectionBuilder.WriteString("\n")
+	valuesSectionBuilder.WriteString("\n### {{ .SectionName }}\n")
```
Comment (mchnwzrd) [inline pkg/document/template.go:214]: Would it be possible to add another `\n` here to ensure compatibility with markdownlint rule `MD022/blanks-around-headings/blanks-around-headers`?

**Comment by norwoodj** [comment]
Sorry for my slowness getting back to this. I've merged. Thank you very much for the very cool contribution

**Pull request merged**